### DISCUSSION TITLE
How to implement cache/memoize

### DISCUSSION THREAD

**Discussion by darky** [gleam-lang/gleam#2789] [Questions & support] [answered]
Hi, congrats with V1! :tada: 
How to implement cache/memoize pattern in Gleam for any argument type?

For example, JavaScript:

```javascript
// Map can use arbitrary types for keys: Objects, Arrays, Functions and so on
const cache = new Map();

const fnUsingCache = (arg) => {
  if (cache.has(arg)) {
    return cache.get(arg)
  }
  
  // resp calculation and cache population
}
```

For example, Golang:

```golang
// Map can use pointers as keys
var cache = map[interface{}]interface{}

func FnUsingCache[T, R Any](arg *T) R {
  if cache[arg] != nil {
    return cache[arg]
  }

  // resp calculation and cache population
}
```

**Comment by lpil** [comment]
Hello! In a function language you pass the cache around as we make state explicit.

**Reply by darky** [reply]
Ok, maybe. But how to implement state for composite types like Records or Functions, which should be used as cache keys?
In JavaScript all reference types can be used as `Map` keys.
In Golang you can get pointer for `Struct` or `Function` and use it as `map` key
In JVM (Clojure) you can get `hash` for anything and use it `map` key
What about Gleam? I found only `Dict` but it uses strings for keys, which very limited.

**Reply by lpil** [reply, marked as answer]
Dict does not use strings for keys, you can use anything.

**Reply by darky** [reply]
Sorry, my bad investigation, thanks!

**Reply by lpil** [reply]
No prob!!
